php/src/Services/Security/SecurityService.php

```diff
@@ -383,30 +388,93 @@ public function checkLoggedInObjectAccess($object, $accessMode = self::ACCESS_RE
-                    } else {
-
-                        $access = $access && $this->getLoggedInScopePrivileges($scopeAccess->getScope(),
-                                $scopeId);
+
+                $objectScopeAccesses = $this->scopeManager->generateObjectScopeAccesses($object, $scopeAccess->getScope());
+
+                // Only continue if at least one object scope access detected
+                if (sizeof($objectScopeAccesses)) {
+
+                    // Now loop through the detected object scope accesses and calculate
+                    foreach ($objectScopeAccesses as $objectScopeAccess) {
+                        $scopeId = $objectScopeAccess->getRecipientPrimaryKey();
+                        $accessGroup = $objectScopeAccess->getAccessGroup();
+
+                        // Ensure access group granted defined
+                        if (!isset($accessGroupGranted[$accessGroup])) {
+                            $accessGroupGranted[$accessGroup] = true;
+                        }
+
+                        if ($scopeId === null || $scopeId == -1) {
+                            if ($scopeAccess->getScope() == Role::SCOPE_ACCOUNT)
+                                $accessGroupGranted[$accessGroup] = $accessGroupGranted[$accessGroup] && ($accessMode == self::ACCESS_READ && ($loggedInSecurable || $loggedInAccount));
+                        } else {
+
+                            // Calculate an initial check as to whether this object has been granted the right level of access
+                            $accessModeMatches = ($accessMode == self::ACCESS_READ) || (($accessMode == self::ACCESS_WRITE) && $objectScopeAccess->getWriteAccess())
+                                || (($accessMode == self::ACCESS_GRANT) && $objectScopeAccess->getGrantAccess());
+
+                            // Compare with logged in privileges if we are accessing non active scopes
+                            // or the scope id matches
+                            if ($accessModeMatches && ($classInspector->hasClassAttribute(AccessNonActiveScopes::class)
+                                    || (!$scopeAccess->getActiveScopeValue() ||
+                                        ($scopeId == $scopeAccess->getActiveScopeValue())))) {
+                                $accessGroupGranted[$accessGroup] = $accessGroupGranted[$accessGroup] && $this->getLoggedInScopePrivileges($scopeAccess->getScope(),
+                                        $scopeId);
+                            } // Otherwise assume false for this
+                            else {
+                                $accessGroupGranted[$accessGroup] = false;
+                            }
+                        }
+
                     }
+
                 }
-                if (!$access)
-                    break;
+
+
             }
 
         }
 
-        return $access;
+        // Return at least one true
+        return sizeof($accessGroupGranted) == 0 || in_array(true, $accessGroupGranted);
```

Comment:
! Why is this boolean or? If accessGroupGranted is empty why does the user have permission?

Reply:
This is to support previous behaviour where there are no access groups in play.  I suspect if it gets this far it won't have been vetoed